File: go/bst/bst.go
package main

import "fmt"

type Node struct {
	Data  int
	Left  *Node
	Right *Node
}

func Insert(root *Node, data int) *Node {
	if root == nil {
		return &Node{Data: data, Left: nil, Right: nil}
	}

	if data < root.Data {
		root.Left = Insert(root.Left, data)
	} else if data > root.Data {
		root.Right = Insert(root.Right, data)
	}

	return root
}

func FindMin(root *Node) *Node {
	for root.Left != nil {
		root = root.Left
	}
	return root
}

func DeleteNode(root *Node, data int) *Node {
	if root == nil {
		return nil
	}

	if data < root.Data {
		root.Left = DeleteNode(root.Left, data)
	} else if data > root.Data {
		root.Right = DeleteNode(root.Right, data)
	} else {
		if root.Left == nil {
			return root.Right
		} else if root.Right == nil {
			return root.Left
		}

		temp := FindMin(root.Right)
		root.Data = temp.Data
		root.Right = DeleteNode(root.Right, temp.Data)
	}
	return root
}

func Search(root *Node, data int) bool {
	if root == nil {
		return false
	}

	if data == root.Data {
		return true
	} else if data < root.Data {
		return Search(root.Left, data)
	} else {
		return Search(root.Right, data)
	}
}

func InorderTraversal(root *Node) {
	if root != nil {
		InorderTraversal(root.Left)
		fmt.Printf("%d ", root.Data)
		InorderTraversal(root.Right)
	}
}

func PreorderTraversal(root *Node) {
	if root != nil {
		fmt.Printf("%d ", root.Data)
		PreorderTraversal(root.Left)
		PreorderTraversal(root.Right)
	}
}

func PostorderTraversal(root *Node) {
	if root != nil {
		PostorderTraversal(root.Left)
		PostorderTraversal(root.Right)
		fmt.Printf("%d ", root.Data)
	}
}

File: cpp/bst/src/bst.cpp
#include "bst.h"

BST::~BST() {
    freeMemory(root);
}

void BST::freeMemory(Node* node) {
    if (node != nullptr) {
        freeMemory(node->left);
        freeMemory(node->right);
        delete node;
    }
}

void BST::insert(int data) {
    root = insertRecursive(root, data);
}

Node* BST::insertRecursive(Node* root, int data) {
    if (root == nullptr) {
        return new Node(data);
    }

    if (data < root->data) {
        root->left = insertRecursive(root->left, data);
    } else if (data > root->data) {
        root->right = insertRecursive(root->right, data);
    }
    return root;
}

bool BST::search(int data) const {
    Node* current = root;
    while (current != nullptr) {
        if (data == current->data) {
            return true;
        } else if (data < current->data) {
            current = current->left;
        } else {
            current = current->right;
        }
    }
    return false;
}

void BST::remove(int data) {
    root = deleteNodeRecursive(root, data);
}

Node* BST::findMin(Node* root) const {
    while (root && root->left != nullptr) {
        root = root->left;
    }
    return root;
}

Node* BST::deleteNodeRecursive(Node* root, int data) {
    if (root == nullptr) {
        return nullptr;
    }

    if (data < root->data) {
        root->left = deleteNodeRecursive(root->left, data);
    } else if (data > root->data) {
        root->right = deleteNodeRecursive(root->right, data);
    } else {
        if (root->left == nullptr) {
            Node* temp = root->right;
            delete root;
            return temp;
        } else if (root->right == nullptr) {
            Node* temp = root->left;
            delete root;
            return temp;
        }

        Node* temp = findMin(root->right);
        root->data = temp->data;
        root->right = deleteNodeRecursive(root->right, temp->data);
    }
    return root;
}

void BST::inorderTraversal() const {
    inorderTraversalRecursive(root);
    std::cout << std::endl;
}

void BST::inorderTraversalRecursive(Node* root) const {
    if (root != nullptr) {
        inorderTraversalRecursive(root->left);
        std::cout << root->data << " ";
        inorderTraversalRecursive(root->right);
    }
}

void BST::preorderTraversal() const {
    preorderTraversalRecursive(root);
    std::cout << std::endl;
}

void BST::preorderTraversalRecursive(Node* root) const {
    if (root != nullptr) {
        std::cout << root->data << " ";
        preorderTraversalRecursive(root->left);
        preorderTraversalRecursive(root->right);
    }
}

void BST::postorderTraversal() const {
    postorderTraversalRecursive(root);
    std::cout << std::endl;
}

void BST::postorderTraversalRecursive(Node* root) const {
    if (root != nullptr) {
        postorderTraversalRecursive(root->left);
        postorderTraversalRecursive(root->right);
        std::cout << root->data << " ";
    }
}